File: str-comb/scomb.js
//////////////////////////////
// scomb.js
function newGap(gap){
    gap /= 1.3;
    gap = gap|0;
    if(gap == 9 || gap == 10) return 11;
    if(gap < 1) return 1;
    return gap;
}
function combSort(a){
    var len = a.length;
    var gap = len;
    var temp;
    var swapped;
    do {
        swapped = false;
        gap = newGap(gap);
        for(var i=0; i < len-gap; ++i) {
            if(a[i] > a[i+gap]) {
                swapped = true;
                temp = a[i];
                a[i] = a[i+gap];
                a[i+gap] = temp;
            }
        }
    } while(gap > 1 || swapped);
}
var N = 10000000;
var arr = new Array(N);
for(var z=0;z<N;++z) arr[z] = '' + (N-z);    
combSort(arr);
for(var z=1;z<N;++z) if(arr[z]<arr[z-1]) console.log("!");

File: str-comb/scomb.scala
//////////////////////////////
// scomb.scala
object scomb {
    def newGap(gap: Int): Int = {
        var ngap : Int = (gap.toDouble / 1.3).toInt;
        if(ngap == 9 || ngap == 10) return 11;
        if(ngap < 1) return 1;
        return ngap;
    }
    def combSort(a: Array[String]) : Array[String] = {
        var res : Array[String] = a;
        val len = a.length;
        var gap = len;
        var temp : String = "";
        var swapped : Boolean = false;
        do {
            swapped = false;
            gap = newGap(gap);
            for(i <- 0 until len-gap if res(i) > res(i+gap)) {
                swapped = true;
                temp = res(i);
                res(i) = res(i+gap);
                res(i+gap) = temp;
            }
        } while(gap > 1 || swapped);
        return res
    }
    val N = 10000000;
    def main(args: Array[String]) {
        var arr : Array[String] = new Array[String](N)
        for(z <- 0 until N) arr(z) = (N-z).toString;   
        arr = combSort(arr); 
        for(z <- 1 until N if arr(z)<arr(z-1)) print("!");
    }
}
